### PR TITLE
Ignore empty lines in indent filter #681

### PR DESCRIPTION
**Pull request by jkfran**
Ignore empty lines in indent filter #681

### CODE REVIEW

**Comment by ThiefMaster** [comment]
Could you please rebase the PR? Generally it's a good idea to use separate branches based off upstream/master for PRs instead of doing the changes in your own master branch.

While this can be done when merging a PR, sometimes people merge using a merge commit instead of rebase/squash and in that case the extra noise would remain in the upstream history forever.

File: jinja2/filters.py [hunk 1]
```diff
@@ -457,7 +457,9 @@ def do_indent(s, width=4, indentfirst=False):
             indent by two spaces and indent the first line too.
     """
     indention = u' ' * width
-    rv = (u'\n' + indention).join(s.splitlines())
+    lines = s.splitlines()
+    rv = lines.pop(0) + u'\n'
+    rv += u'\n'.join([indention + l if l else l for l in lines])
```
Comment (ThiefMaster) [inline jinja2/filters.py]: lowercase-L is not a good choice (and actually disallowed as per PEP8, just like uppercase-i)

**Comment by jkfran** [comment]
Hello @ThiefMaster thank you!
I'm sorry for the merge commits. What do you think now?

File: jinja2/filters.py [hunk 2]
```diff
@@ -457,7 +457,9 @@ def do_indent(s, width=4, indentfirst=False):
             indent by two spaces and indent the first line too.
     """
     indention = u' ' * width
-    rv = (u'\n' + indention).join(s.splitlines())
+    lines = s.splitlines()
+    rv = lines.pop(0) + u'\n'
+    rv += u'\n'.join([indention + line if line else line for line in lines])
```
Comment (ThiefMaster) [inline jinja2/filters.py]: the `[]` isn't needed since there's only one argument to `join()`

Comment (jkfran) [inline jinja2/filters.py]: @ThiefMaster well it's Generator Expressions vs. List Comprehension, I was reading and you are right for these case Generator Expression is better.
http://stackoverflow.com/questions/47789/generator-expressions-vs-list-comprehension

Comment (jkfran) [inline jinja2/filters.py]: @ThiefMaster well it's Generator Expressions vs. List Comprehension, I was reading and you are right for these case Generator Expression is better.
http://stackoverflow.com/questions/47789/generator-expressions-vs-list-comprehension

**Comment by jkfran** [comment]
Hello @ThiefMaster what do you think?

**Comment by jkfran** [comment]
@ThiefMaster Excuse me but I need to do something? Close de PR or something? When it will be merged?
(Newbie sorry) :)

**Comment by ThiefMaster** [comment]
Could you add an entry to the changelog file? The PR looks good so far (not a fan of the all-lowercase-no-underscore naming of the argument but the existing one uses the same scheme it's consistent).

**Comment by jkfran** [comment]
@ThiefMaster Thank you! Everything is alright?

**Comment by jkfran** [comment]
@ThiefMaster

**Comment by davidism** [comment]
I'd like to see a more thorough test suite, it seems like we could be missing a lot of corner cases with this. Also, be sure to test the different options independently, don't just add the new behavior on to an existing test.

File: jinja2/filters.py [hunk 3]
```diff
@@ -457,7 +457,12 @@ def do_indent(s, width=4, indentfirst=False):
             indent by two spaces and indent the first line too.
     """
     indention = u' ' * width
-    rv = (u'\n' + indention).join(s.splitlines())
+    if indentblanklines:
+        rv = (u'\n' + indention).join(s.splitlines())
+    else:
+        lines = s.splitlines()
+        rv = lines.pop(0) + u'\n'
```
Comment (davidism) [inline jinja2/filters.py]: This adds a trailing newline if there is only one line, unlike `indentblanklines=True`.

File: jinja2/filters.py [hunk 4]
```diff
@@ -457,7 +457,12 @@ def do_indent(s, width=4, indentfirst=False):
...
+        rv += u'\n'.join(indention + line if line else line for line in lines)
```
Comment (davidism) [inline jinja2/filters.py]: This removes trailing blank lines, unlike `indentblanklines=False`.

File: tests/test_filters.py [hunk 5]
```diff
@@ -126,11 +126,12 @@ def test_format(self, env):
         assert out == 'a|b'
 
     def test_indent(self, env):
-        tmpl = env.from_string('{{ foo|indent(2) }}|{{ foo|indent(2, true) }}')
-        text = '\n'.join([' '.join(['foo', 'bar'] * 2)] * 2)
+        tmpl = env.from_string('{{ foo|indent(2) }}|'
+                               '{{ foo|indent(2, true, true) }}')
```
Comment (davidism) [inline tests/test_filters.py]: Let's split each case into separate renders rather than separating them with `|`.

File: tests/test_filters.py [hunk 6]
```diff
@@ -126,11 +126,12 @@ def test_format(self, env):
...
+        text = '\n'.join(['foo bar', '', 'foo baz'] * 2)
         out = tmpl.render(foo=text)
-        assert out == ('foo bar foo bar\n  foo bar foo bar|  '
-                       'foo bar foo bar\n  foo bar foo bar')
+        assert out == ('foo bar\n\n  foo baz\n  foo bar\n\n  foo baz|  '
+                       'foo bar\n  \n  foo baz\n  foo bar\n  \n  foo baz')
```
Comment (davidism) [inline tests/test_filters.py]: This doesn't cover `indent(2, False, True)` or `indent(2, True, False)`.

File: jinja2/filters.py [hunk 7]
```diff
@@ -445,7 +445,7 @@ def do_urlize(eval_ctx, value, trim_url_limit=None, nofollow=False,
     return rv
 
 
-def do_indent(s, width=4, indentfirst=False):
+def do_indent(s, width=4, indentfirst=False, indentblanklines=False):
```
Comment (davidism) [inline jinja2/filters.py]: I know @ThiefMaster said it's consistent with `indentfirst`, but I'd rather just break consistency and use the easier to read `indent_blank_lines`.

Comment (davidism) [inline jinja2/filters.py]: If you want to, also change the signature to `do_indent, s, width=4, indent_first=False, indent_blank_lines=False, indentfirst=None)` and raise a deprecation warning if `indentfirst` is not None. Not required though, I can follow up with this after merging.
